### PR TITLE
Graphical Set

### PR DESCRIPTION
**Pull request by LouiseKlodt**
- replace attributed strings with custom CardView using drawing with UIBezier path
- add gestures (swipe, tap, rotate)
- reorganise cards on screen when more cards dealt / removed 
- reshuffle cards option
- deal more cards until deck empty

There is room for improvement: i.e. the tap gesture recognisers are rebuilt each time the card views change, but it's a fully working app.

### CODE REVIEW

File: SoloSet/CardView.swift [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        shape: .diamonds,
+        shading: .solid,
+        color: CustomColors.swBlack) { didSet { setNeedsDisplay() } }
+
+    override func draw(_ rect: CGRect) {
+        
+        // helper functions
+        // ----------------
+        
+        func diamond(_ p: UIBezierPath, _ cx: CGFloat, _ cy: CGFloat, _ w: CGFloat, _ h: CGFloat) {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: by general I'm not a big fan of nested functions. While they are perfectly legit in swift, they make the code hard to read. I suggest using private functions instead

File: SoloSet/CardView.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            p.move(to: CGPoint(x: cx, y: cy - h))
+            p.addLine(to: CGPoint(x: cx - w, y: cy))
+            p.addLine(to: CGPoint(x: cx, y: cy + h))
+            p.addLine(to: CGPoint(x: cx + w, y: cy))
+            p.close()
+        }
+        
+        func squiggles(_ p: UIBezierPath, _ centreX: CGFloat, _ centreY: CGFloat, _ w: CGFloat, _ h: CGFloat ) {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: Would you do the same job with CAShapeLayer? The principle is the same and most of the code too. You'll have to do some researches to find out how to do these shapes with CAShapeLayer. You can start searching for tutorials on how to use them.

File: SoloSet/CardView.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            let ctr2 = CGPoint(x: centreX + (xOff * w), y: centreY - (yOff * h))
+            let ctr3 = CGPoint(x: centreX + (xOff * w), y: centreY + (yOff * h))
+            let ctr4 = CGPoint(x: centreX - (xOff * w), y: centreY + (yOff * h))
+            p.move(to: start)
+            p.addCurve(to: end, controlPoint1: ctr1 , controlPoint2: ctr2)
+            p.move(to: end)
+            p.addCurve(to: start, controlPoint1: ctr3 , controlPoint2: ctr4)
+        }
+        
+        func yPositions(_ n: Int) -> [CGFloat] {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: what n represents?

File: SoloSet/CardView.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            case 2: cYs = [1/4 * maxY, 3/4 * maxY]
+            case 3: cYs = [1/5 * maxY, 1/2 * maxY, 4/5 * maxY]
+            default: cYs = []
+                
+            }
+            return cYs
+        }
+        
+        func scaleFactor(_ n: Int) -> CGFloat {
+            var k: CGFloat
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: this could be 
`return 1.0/CGFloat(n)`

File: SoloSet/CardView.swift [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            case 2: k = 1/2
+            case 3: k = 1/3
+            default: k = 1.0
+            }
+            return k
+        }
+        
+        func drawShading(_ path: UIBezierPath, _ color: UIColor, _ shading: Shading) {
+            switch shading {
+            case .outlined: color.setStroke(); path.lineWidth = 1.6;  path.stroke()
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: in swift we prefer to separate expressions with new lines. they make the code more readable

File: SoloSet/CardView.swift [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            switch n {
+            case 1: k = 1.0
+            case 2: k = 1/2
+            case 3: k = 1/3
+            default: k = 1.0
+            }
+            return k
+        }
+        
+        func drawShading(_ path: UIBezierPath, _ color: UIColor, _ shading: Shading) {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: This could be a UIBezierPath extension. Others might benefit from this function.

File: SoloSet/CardView.swift [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            switch shading {
+            case .outlined: color.setStroke(); path.lineWidth = 1.6;  path.stroke()
+            case .solid: color.setFill(); path.fill()
+            case .striped: drawStripes(path, color)
+            }
+        }
+        
+        func drawStripes(_ path: UIBezierPath, _ color: UIColor) {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: same here. This could be in an UIBezierPath extension

File: SoloSet/CardView.swift [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            let stripeDist = Double(b.maxX / Constants.stripeDistFactor)
+            for x in stride(from: Double(b.minX), to: Double(b.maxX), by: stripeDist) {
+                path.move(to: CGPoint(x: x, y: Double(b.minY)))
+                path.addLine(to: CGPoint(x: x, y: Double(b.maxY)))
+                path.stroke()
+            }
+            context.restoreGState()
+        }
+        
+        func drawShapes(width w: CGFloat, height h: CGFloat, shape: Shape, shading: Shading, color: UIColor, n: Int) {
```
Comment (gringoireDM) [inline SoloSet/CardView.swift]: Would you try to extract this function in a Shape extension?

File: SoloSet/SoloSet.swift [hunk 9]
```diff
@@ -40,21 +38,24 @@ struct SoloSet {
...
             if isSet() {
+                set = true
                 if deck.count >= cardsInSet {
                     replace(at: selected)
                 } else {
-                    for indx in selected {
-                        cards[indx].isDone = true
-                        cards[indx].isSelected = false
+                    let sortedSelected = selected.sorted().reversed()
+                    for indx in sortedSelected {
```
Comment (gringoireDM) [inline SoloSet/SoloSet.swift]: you could do `selected.sorted().reversed().forEach(cards.remove)`

File: SoloSet/SoloSet.swift [hunk 10]
```diff
@@ -40,21 +38,24 @@ struct SoloSet {
     }
     
     
-    mutating func updateCards() {
+    mutating func updateCards() -> Bool {
+        var set = false
         if selected.count == cardsInSet {
```
Comment (gringoireDM) [inline SoloSet/SoloSet.swift]: Is there any way you could reduce the amount of nested `if`s?

File: SoloSet/CardView.swift (hunk 2 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: OK, looking into that.

File: SoloSet/CardView.swift (hunk 3 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: It's the number of items to draw on the card. Added a more descriptive name.

File: SoloSet/CardView.swift (hunk 4 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: Cheers; removed that function.. thanks for spotting this.

File: SoloSet/CardView.swift (hunk 5 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: Got it; changed that.

File: SoloSet/CardView.swift (hunk 6 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: OK, changed.

File: SoloSet/CardView.swift (hunk 7 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: OK changed.

File: SoloSet/SoloSet.swift (hunk 9 above)
Comment (LouiseKlodt) [reply SoloSet/SoloSet.swift]: Tried that, but it throws an error. `forEach` returns Void, but `remove` returns the item removed.

File: SoloSet/SoloSet.swift (hunk 10 above)
Comment (LouiseKlodt) [reply SoloSet/SoloSet.swift]: I don't see a way.. am I missing something?

File: SoloSet/CardView.swift (hunk 1 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: OK, changed that.

File: SoloSet/CardView.swift (hunk 8 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: Not sure what you mean by that? You mean add the function to the Shape type? The type is defined in the model, and I want to separate drawing from the model.

File: SoloSet/CardView.swift (hunk 2 above)
Comment (LouiseKlodt) [reply SoloSet/CardView.swift]: I could only find examples where `CAShapeLayer` was used in combination with a path, e.g. `Bezierpath` as I did it; see for example https://www.calayer.com/core-animation/2016/05/22/cashapelayer-in-depth.html. The Apple doc is very sparse. Could you point me to code examples where `CAShapeLayer` is used to draw the actual shapes?

**Pull request closed without merging**